### PR TITLE
[GPDB_12_MERGE_FIXME] support toast_tuple_target

### PR DESCRIPTION
**Pull request by interma**
Original logic and improvements:
1. heap table
> it ignores the toast_tuple_target setting of reloptions, just using TOAST_TUPLE_TARGET

Improve it to use reloptions, so the user's setting value can take effect (e.g. `alter table T set (toast_tuple_target=5000)` ).

2. ao table
> set it to toast_tuple_target param and all callers pass TOAST_TUPLE_TARGET

Considering reloptions for AO table is not permitted, using TOAST_TUPLE_TARGET to make code clearly (with correct length: target_length - header_len, same as heap table ).

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [x] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/access/heap/tuptoaster.c [hunk 1]
```diff
@@ -876,13 +876,13 @@ toast_insert_or_update_generic(Relation rel, void *newtup, void *oldtup,
...
 	}
 	else
 	{
-		// GPDB_12_MERGE_FIXME: use RelationGetToastTupleTarget here too?
-		maxDataLen = toast_tuple_target;
-		hoff = -1; /* keep compiler quiet about using 'hoff' uninitialized */
+		/* Since reloptions for AO table is not permitted, so using TOAST_TUPLE_TARGET */
+		hoff = sizeof(MemTupleData);
+		hoff = MAXALIGN(hoff);
+		maxDataLen = TOAST_TUPLE_TARGET - hoff;
```
Comment (adam8157) [inline src/backend/access/heap/tuptoaster.c:885]: Shall we use the lower case `toast_tuple_target`? Which is said **dynamic** for AO, checkout `appendonly_insert_init`, I'm not sure.

**Comment by adam8157** [comment]
> Improve it to use reloptions, so the user's setting value can take effect (e.g. alter table T set (toast_tuple_target=5000) ).

How does it take effect? I'm lost here.

File: src/backend/access/heap/tuptoaster.c (hunk 1 above)
Comment (interma) [reply src/backend/access/heap/tuptoaster.c:885]: Thanks @adam8157 ,  I will check the code.

**Comment by interma** [comment]
> > Improve it to use reloptions, so the user's setting value can take effect (e.g. alter table T set (toast_tuple_target=5000) ).
> 
> How does it take effect? I'm lost here.

The previous code is:
```
		maxDataLen = RelationGetToastTupleTarget(rel, TOAST_TUPLE_TARGET) - hoff;
		maxDataLen = TOAST_TUPLE_TARGET - hoff; // <== always use TOAST_TUPLE_TARGET here
```
It ignores the setting of the first line.

File: src/backend/access/heap/tuptoaster.c (hunk 1 above)
Comment (interma) [reply src/backend/access/heap/tuptoaster.c:885]: I tested, user cannot set toast_tuple_target for AO table:
```
demo=# create table test_ao_toast(id int, title varchar, body text) with (appendonly=true, toast_tuple_target=5000) distributed by (id);
psql: ERROR:  unrecognized parameter "toast_tuple_target"
```
And checked code, the default toast_tuple_target of ao is always TOAST_TUPLE_TARGET.
So, I think it's clearer to set TOAST_TUPLE_TARGET here, toast_tuple_target param may be deleted in future (doesn't broke ABI now).

**Comment by adam8157** [comment]
Ah, I see it in `RelationGetToastTupleTarget()`, I must be sleepy that day, sorry.

**Review by adam8157 — APPROVED** [review]

**Pull request merged**